static/js/message_report.js

```diff
@@ -0,0 +1,115 @@
+    if (!$row) {
+        return;
+    }
+
+    // in case this message isn't selected (which happens)
+    message_lists.current.select_id(message.id, {then_scroll: true});
+    const $target = $row.find(".messagebox-content");
+    // compute the top of the report dialog box
+    const scrollDest = message_viewport.$message_pane.scrollTop() + $target.safeOuterHeight() - 40;
+    const width = $row.safeOuterWidth() - 54;
+
+    // add the report dialog HTML:
+    //  - append() is used to absolutely minimize the DOM overhead of this rarely used feature,
+    //    i.e. we don't want *any* extra per-message DOM elements for msgs that aren't reported.
+    //  - remove() called in case the user attempts to call report multiple times.
+    const html = render_report_message();
+    $target.find(".reportmsg-box").remove();
+    $target.append(html);
+    $target.find(".reportmsg-box").css("width", width + "px");
+
+    // nice little animation to gently open the dialog box and then scroll down to it,
+    // drawing the user's eyes to the reporting form.  note: for uncondensed long messages,
+    // this initial animation (N msec) will happen off-screen, which is fine in practice.
+    const $reportmsg = $row.find(".reportmsg-box");
+    $reportmsg.slideDown(600, () => {
+        message_viewport.$message_pane.animate({scrollTop: scrollDest}, 600);
+    });
+    $reportmsg.find("div,label,input[type=radio]").on("click", (e) => {
+        // don't initiate a reply-message, which is the default behavior in zulip
+        e.stopPropagation();
+    });
+    $reportmsg.find(".reportmsg_cancel").on("click", (e) => {
+        $reportmsg.slideUp(400);
+        e.stopPropagation();
+        e.preventDefault();
+    });
+    $reportmsg.find(".reportmsg_submit").on("click", (e) => {
+        e.stopPropagation();
+        e.preventDefault();
+
+        const reason = $reportmsg.find("input[type=radio][name=reportmsg-reason]:checked").val();
+        if (!["spam", "harassment", "inappropriate", "norms", "other"].includes(reason)) {
+            msgreport_alert($t({defaultMessage: "Please select a reason."}));
+            return;
+        }
+        let explanation = $reportmsg.find("textarea[name=reportmsg-explain]").val();
+        if (explanation === undefined) {
+            explanation = "";
+        }
+        if (reason === "other" && explanation === "") {
+            msgreport_alert($t({defaultMessage: "'Other' requires an explanation."}));
+            return;
+        }
+
+        if (message && message.raw_content) {
+            report_message(message, $reportmsg, reason, explanation);
+            return;
+        }
+        channel.get({
+            url: "/json/messages/" + message.id,
+            idempotent: true,
+            success(data) {
+                message.raw_content = data.raw_content;
+                report_message(message, $reportmsg, reason, explanation);
+            },
```

Comment:
We probably need something to happen if an error occurs in this API call, which should be rare. 

That said, one theory could be to do the API call to fetch raw content when you start the report message experience, rather than at the end. You can crib a bit off the message edit code path -- I'm ~50% sure we have existing code to check if `raw_content` is available, and if not fetch it and then continue.

Reply:
NOCHG - no longer relevant!  As per previous review, I moved the quoting logic to the server and deleted this code, which means the client no long needs the message content to report a message. Double win!